src/components/Orientation training/Timer/Timer.tsx

```diff
@@ -0,0 +1,53 @@
+import { useState, useEffect, useRef } from "react";
+import TimerIcon from "@mui/icons-material/Timer";
+import { Box, Typography } from "@mui/material";
+import * as style from "./Timer.style";
+
+const Timer = ({
+  isTurnedOn,
+  onTimerEnd,
+}: {
+  isTurnedOn: boolean;
+  onTimerEnd: () => void;
+}) => {
+  const [totalSeconds, setTotalSeconds] = useState(30);
+  const intervalRef = useRef<NodeJS.Timeout | null>(null);
+
+  useEffect(() => {
+    if (isTurnedOn) {
+      if (!intervalRef.current) {
+        intervalRef.current = setInterval(() => {
+          setTotalSeconds((prev) => {
+            if (prev === 0) {
+              clearInterval(intervalRef.current!);
+              intervalRef.current = null;
+              onTimerEnd();
+              return 0;
+            }
+            return prev - 1;
+          });
+        }, 1000);
+      }
+    } else {
+      setTotalSeconds(30);
+    }
+
+    return () => {
```

Comment:
```suggestion
  useEffect(() => {
    if (!isTurnedOn) {
      setTotalSeconds(30);
      return;
    }

    if (intervalRef.current) return;

    intervalRef.current = setInterval(() => {
      setTotalSeconds((prev) => {
        if (prev > 0) {
          return prev - 1;
        }
        
        clearInterval(intervalRef.current!);
        intervalRef.current = null;
        onTimerEnd();
        return 0;
      });
    }, 1000);
```

Ale przede wszystkim może nawet w osobnym branchu zrób hook, który użyjesz tu i w StopWatch